types/types.go

```diff
@@ -816,22 +781,203 @@ func TypeOf(x interface{}) Type {
 	panic("unreachable")
 }
 
+// FormatTypeKeyElide returns a string represenation of the given type
+// with the fields elided
+func FormatTypeKeyElide(t Type) string {
+	switch t := t.(type) {
+	case *Object:
+		prefix := "object"
+		if len(t.static) > 0 {
+			prefix += "<...>"
+		}
+		if t.dynamic != nil {
+			prefix += "[...]"
+		}
+		return prefix
+	case *Array:
+		prefix := "array"
+		if len(t.static) > 0 {
+			prefix += "<...>"
+		}
+		if t.dynamic != nil {
+			prefix += "[...]"
+		}
+		return prefix
+	case *Set:
+		return "set[...]"
+	case Any:
+		return "any<...>"
+	default:
+		return Sprint(t)
+	}
+}
+
 type typeSlice []Type
 
 func (s typeSlice) Less(i, j int) bool { return Compare(s[i], s[j]) < 0 }
 func (s typeSlice) Swap(i, j int)      { x := s[i]; s[i] = s[j]; s[j] = x }
 func (s typeSlice) Len() int           { return len(s) }
 
-func typeSliceCompare(a, b []Type) int {
+func typeArrayCompare(a, b Type, args ...*[]Type) int {
```

Comment:
In general, I would avoid use of variable length argument lists. In some cases they make the API more readable, but in this case, they're being used to reduce duplicated logic. However, we still need two code paths below.